shotover/src/codec/cassandra.rs

```diff
@@ -674,65 +706,115 @@ impl CassandraEncoder {
+                            let header_start = dst.len();
+
+                            dst.extend_from_slice(&[0, 0, 0, 0, 0, 0]);
+                            let payload_start = dst.len();
 
-                        if true {
-                            // TODO if self_contained
+                            let mut payload_len = envelope_bytes.len();
+                            dst.put(envelope_bytes);
+
+                            // self contained flag
                             payload_len |= 1 << 17;
-                        }
 
-                        // add header length & header crc
-                        let payload_len = &payload_len.to_le_bytes()[..3];
-                        dst[header_start..header_start + 3].copy_from_slice(payload_len);
-                        dst[header_start + 3..header_start + 6]
-                            .copy_from_slice(&crc24(payload_len).to_le_bytes()[..3]);
+                            // add header length & header crc
+                            let payload_len = &payload_len.to_le_bytes()[..3];
+                            dst[header_start..header_start + 3].copy_from_slice(payload_len);
+                            dst[header_start + 3..header_start + 6]
+                                .copy_from_slice(&crc24(payload_len).to_le_bytes()[..3]);
 
-                        // add payload crc
-                        dst.extend_from_slice(&crc32(&dst[payload_start..]).to_le_bytes());
+                            // add payload crc
+                            dst.extend_from_slice(&crc32(&dst[payload_start..]).to_le_bytes());
+                        }
                     }
                     Compression::Lz4 => {
-                        let header_start = dst.len();
+                        let mut envelope_bytes = self.encode_envelope(m, Compression::None)?;
 
-                        dst.extend_from_slice(&[0, 0, 0, 0, 0, 0, 0, 0]);
-                        let payload_start = dst.len();
+                        if get_maximum_output_size(envelope_bytes.len()) > PAYLOAD_SIZE_LIMIT {
+                            while !envelope_bytes.is_empty() {
+                                let header_start = dst.len();
 
-                        // TODO we should not be encoding small frames
-                        let (uncompressed_len, compressed_len) =
-                            self.encode_compressed_payload(dst, m, payload_start)?;
+                                dst.extend_from_slice(&[0, 0, 0, 0, 0, 0, 0, 0]);
+                                let payload_start = dst.len();
 
-                        if compressed_len > PAYLOAD_SIZE_LIMIT {
-                            todo!("non self contained frames not yet implemented.")
-                        }
+                                let payload_bytes = envelope_bytes
+                                    .split_to(envelope_bytes.len().min(PAYLOAD_SIZE_LIMIT - 1));
```

Comment:
If we take a section of payload of length PAYLOAD_SIZE_LIMIT and compress it, the final compressed output may be larger than PAYLOAD_SIZE_LIMIT.

If we apply some algebra to get_maximum_output_size we can derive:
```rust
fn maximum_input_size(output_size: usize) -> usize {
    ((output_size - 20) as f32 / 1.1) as usize
}
```

And then we can do `maximum_input_size(PAYLOAD_SIZE_LIMIT)` to get the maximum size we can use here.

Reply:
`encode_compressed_payload_into_buffer` checks if it's larger than `uncompressed_len` which is going to be at maximum `PAYLOAD_SIZE_LIMIT -1`

Reply:
Oh, I missed that!
Yeah, that looks good, maybe just add a comment to make it clear that logic is not just for effeciency but also important for correctness?

```diff
diff --git a/shotover/src/codec/cassandra.rs b/shotover/src/codec/cassandra.rs
index 57bf38f2..7fb4b3bd 100644
--- a/shotover/src/codec/cassandra.rs
+++ b/shotover/src/codec/cassandra.rs
@@ -841,6 +841,8 @@ impl CassandraEncoder {
         let mut compressed_len = compress_into(bytes, &mut dst[payload_start..])?;
 
         // fallback to uncompressed data if its more efficient
+        // this is also important for correctness, if the compressed len is larger than the uncompressed len,
+        // the compressed payload could overflow the max message size that was checked against the uncompressed size
         if compressed_len > uncompressed_len {
             dst[payload_start..(payload_start + uncompressed_len)]
                 .copy_from_slice(&bytes[..uncompressed_len]);
```